### PR TITLE
added support for sending telemetry to eventhubs

### PR DESCRIPTION
**Pull request by martin-weber**
- added EventHubConnectionString command line parameter to IotTelemetrySimulator
- moved IoT Hub and Event Hub specific code out of SimulatedDevice into implementations of ISender to make SimulatedDevices agnostic of the message sink
- moved IotHub DeviceClient out of Factory int ISender implementation of the corresponding message sink

### CODE REVIEW

File: src/IotTelemetrySimulator/EventHubSender.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            const int MaxAttempts = 3;
+            var msg = CreateMessage();
+
+            for (var attempt = 1; attempt <= MaxAttempts; ++attempt)
+            {
+                try
+                {
+                    await eventHubClient.SendAsync(msg);
+                    stats.IncrementMessageSent();
```
Comment (fbeltrao) [inline src/IotTelemetrySimulator/EventHubSender.cs]: We are repeating the `stats.IncrementMessageSent()` and `stats.IncrementSendTelemetryErrors()` in both senders. Wonder if we can share this code in a common class (i.e. composition pattern)

File: src/IotTelemetrySimulator/DefaultDeviceSimulatorFactory.cs [hunk 2]
```diff
@@ -1,26 +1,25 @@
...
+        }
+
+        private ISender GetSender(string deviceId, RunnerConfiguration config)
+        {
+            if (!string.IsNullOrEmpty(config.IotHubConnectionString))
+                return new IotHubSender(deviceId, config);
+
+
+            if (!string.IsNullOrEmpty(config.EventHubConnectionString))
+                return new EventHubSender(deviceId, config);
```
Comment (fbeltrao) [inline src/IotTelemetrySimulator/DefaultDeviceSimulatorFactory.cs]: We should reuse the EventHubClient here

**Review by fbeltrao — COMMENTED** [review]
@martin-weber Thanks for contribution. Sorry for the delay. Finally had the time to look at it.

File: src/IotTelemetrySimulator/EventHubSender.cs (hunk 1 above)
Comment (martin-weber) [reply src/IotTelemetrySimulator/EventHubSender.cs]: I added an abstract base class SenderBase to reduce code duplication

File: src/IotTelemetrySimulator/DefaultDeviceSimulatorFactory.cs (hunk 2 above)
Comment (martin-weber) [reply src/IotTelemetrySimulator/DefaultDeviceSimulatorFactory.cs]: Agree. Done

**Comment by fbeltrao** [comment]
@martin-weber could you please resolve the conflicts?

**Comment by martin-weber** [comment]
Refactored code to have a common SenderBase which reduced duplicated code to a minimum.

File: src/IotTelemetrySimulator/SenderBase.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+            for (var attempt = 1; attempt <= MaxAttempts; ++attempt)
+            {
+                try
+                {
+                    await SendAsync(msg, cancellationToken);
+                    stats.IncrementMessageSent();
+                    break;
+                }
+                catch (TCommunicationException)
```
Comment (fbeltrao) [inline src/IotTelemetrySimulator/SenderBase.cs]: Maybe if we define an virtual/abstract method called `bool IsTransientException(Exception)` we can remove the TCommunicationException. I think it is a bit easier to follow.

File: src/IotTelemetrySimulator/SenderBase.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace IotTelemetrySimulator
+{
+    public abstract class SenderBase<TMessage, TCommunicationException> : ISender
```
Comment (fbeltrao) [inline src/IotTelemetrySimulator/SenderBase.cs]: Nice one!

File: src/IotTelemetrySimulator/SenderBase.cs (hunk 3 above)
Comment (martin-weber) [reply src/IotTelemetrySimulator/SenderBase.cs]: Good idea.

**Comment by martin-weber** [comment]
added unit tests for SenderBase

**Review by fbeltrao — APPROVED** [review]

**Pull request merged**